File: DifferentialGeometry.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D


class DifferentialGeometry:
    def __init__(self, n, coordinate, X, g, connection):
        self.n = n
        self.coordinate = coordinate
        self.X = X
        self.g = g
        self.connection = connection
        self.Riemann_curvature_tensor = None
        self.Riemann_curvature_metric_tensor = None
        self.Ricci_tensor = None
        self.Scalar_curvature = None
        self.Gauss_curvature = None
        self.FaceColor = 'blue'
        self.alpha = 1.0
        self.scale = 1.0

    def geodesic(self, interval, Init, interval_num):
        t = np.linspace(interval[0], interval[1], interval_num)
        X = np.zeros((interval_num, self.n))
        X[0] = Init
        for i in range(interval_num - 1):
            V = np.zeros(self.n)
            for j in range(self.n):
                for k in range(self.n):
                    V[j] += self.connection[k][j][j](X[i]) * self.X[k].diff(self.coordinate[j]).subs(
                        [(self.coordinate[l], X[i][l]) for l in range(self.n)])
                V[j] = V[j].subs([(self.coordinate[l], X[i][l]) for l in range(self.n)])
            X[i + 1] = X[i] + t[1] * V
        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
        ax.plot(X[:, 0], X[:, 1], X[:, 2])
        plt.show()

    def parallel_transport(self, u, interval, Init, interval_num):
        t = np.linspace(interval[0], interval[1], interval_num)
        X = np.zeros((interval_num, self.n))
        X[0] = Init
        V = np.zeros(self.n)
        V[0] = 1.0
        for i in range(interval_num - 1):
            for j in range(self.n):
                for k in range(self.n):
                    V[j] += self.connection[k][j][j](X[i]) * V[k] * t[1]
                V[j] = V[j].subs([(self.coordinate[l], X[i][l]) for l in range(self.n)])
            X[i + 1] = X[i] + t[1] * V
        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
        ax.plot(X[:, 0], X[:, 1], X[:, 2])
        plt.show()

    def drawmesh(self, uinterval, vinterval, draw_curvature, interval_num):
        u = np.linspace(uinterval[0], uinterval[1], interval_num)
        v = np.linspace(vinterval[0], vinterval[1], interval_num)
        U, V = np.meshgrid(u, v)
        X = np.zeros((interval_num, interval_num, self.n))
        for i in range(interval_num):
            for j in range(interval_num):
                X[i][j] = [self.X[k].subs([(self.coordinate[l], [U[i][j], V[i][j]]) for l in range(2)]) for k in
                           range(self.n)]
        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
        if draw_curvature:
            if self.n != 3:
                raise ValueError("Can only draw curvature for 3D surfaces")
            if self.Gauss_curvature is None:
                self.calculate_curvature()
            colors = self.Gauss_curvature.ravel()
            cmap = plt.get_cmap('coolwarm')
            surface = ax.plot_surface(X[:, :, 0], X[:, :, 1], X[:, :, 2], facecolors=cmap(colors), shade=False)
            surface.set_facecolor((0,0,0,0))
            m = cm.ScalarMappable(cmap=cmap)
            m.set_array(colors)
            cbar = plt.colorbar(m)
            cbar.ax.set_ylabel('Gaussian Curvature')
        else:
            facecolor = self.FaceColor
            alpha = self.alpha
            surface = ax.plot_surface(X[:, :, 0], X[:, :, 1], X[:, :, 2], facecolors=facecolor, alpha=alpha)
        ax.set_xlabel(self.coordinate[0])
        ax.set_ylabel(self.coordinate[1])
        ax.set_zlabel(self.coordinate[2])
        ax.set_title('Surface')
        scale = self.scale
        if scale != 1:
            X = np.array(X)
            X[:, :, 2] *= scale
        plt.show()

File: DifferentialGeometry.m
classdef DifferentialGeometry
    properties
        n           % 坐标系的维度
        coordinate  % 表示坐标系的符号数组
        X           % 定义曲面或流形的坐标函数
        g           % 定义度量张量的符号数组
        connection  % 定义联络的符号数组
        R           % Riemann 曲率张量
        Rm          % 与度量张量相联系的 Riemann 曲率张量
        Ric         % Ricci 张量
        R_scalar    % 标量曲率
        K           % 高斯曲率
        FaceColor   % 绘制曲面时的颜色
        Alpha       % 绘制曲面时的透明度
        scale       % 平行移动向量的比例尺
    end
    
    methods
        function obj = DifferentialGeometry(n, coordinate, X, g, connection)
            % 构造函数
            obj.n = n;
            obj.coordinate = coordinate;
            obj.X = X;
            obj.g = g;
            obj.connection = connection;
            
            % 计算曲率和高斯曲率
            R = sym('R%d%d%d%d', [n n n n]);
            Rm = sym('Rm%d%d%d%d', [n n n n]);
            Ric = sym('Ric%d%d', [n n]);
            R_scalar = sym('R_scalar');
            K = sym('K');
            for i = 1:n
                for j = 1:n
                    for k = 1:n
                        for l = 1:n
                            R(i,j,k,l) = diff(connection(i,j,l), obj.coordinate(k)) - diff(connection(i,j,k), obj.coordinate(l)) ...
                                + connection(i,m,k)*connection(m,j,l) - connection(i,m,l)*connection(m,j,k);
                            Rm(i,j,k,l) = R(i,j,k,l) + g(i,m)*R(m,j,k,l);
                        end
                    end
                end
            end
            for i = 1:n
                for j = 1:n
                    Ric(i,j) = sum(Rm(i,m,j,m), [m 1 n]);
                end
            end
            R_scalar = sum(sum(g.*Ric));
            K = R_scalar / det(g);
            
            obj.R = R;
            obj.Rm = Rm;
            obj.Ric = Ric;
            obj.R_scalar = R_scalar;
            obj.K = K;
            
            % 设置默认绘图参数
            obj.FaceColor = 'red';
            obj.Alpha = 0.5;
            obj.scale = 0.2;
        end
        
        function geodesic(obj, interval, Init, interval_num)
            % 计算并绘制测地线
            syms t real
            eqs = sym('eq%d', [1 obj.n]);
            for i = 1:obj.n
                eqs(i) = diff(obj.X(i), t, 2) + sum(sum(obj.connection(i,j,k)*diff(obj.X(j), t)*diff(obj.X(k), t), [j 1 obj.n]), [k 1 obj.n]);
            end
            [T, Y] = ode45(@(t,y) double(subs(eqs, [sym('t'), obj.coordinate], [t, y])), interval, Init);
    X = zeros(length(T), obj.n);
    for i = 1:obj.n
        X(:,i) = Y(:,i);
    end
    plot3(X(:,1), X(:,2), X(:,3), 'LineWidth', 2);
        end
    end 
end
